// @steipete/summarize/dist/esm/content/link-preview/content/utils.js
import { applyContentBudget, normalizeCandidate, normalizeForPrompt } from './cleaner.js';
import { DEFAULT_CACHE_MODE, DEFAULT_MAX_CONTENT_CHARACTERS, DEFAULT_TIMEOUT_MS, } from './types.js';
const WWW_PREFIX_PATTERN = /^www\./i;
const TRANSCRIPT_LINE_SPLIT_PATTERN = /\r?\n/;
const WORD_SPLIT_PATTERN = /\s+/g;
export function resolveCacheMode(options) {
    return options?.cacheMode ?? DEFAULT_CACHE_MODE;
}
export function resolveMaxCharacters(options) {
    const candidate = options?.maxCharacters;
    if (typeof candidate !== 'number' || !Number.isFinite(candidate) || candidate <= 0) {
        return null;
    }
    if (candidate <= DEFAULT_MAX_CONTENT_CHARACTERS) {
        return DEFAULT_MAX_CONTENT_CHARACTERS;
    }
    return Math.floor(candidate);
}
export function resolveTimeoutMs(options) {
    const candidate = options?.timeoutMs;
    if (typeof candidate !== 'number' || !Number.isFinite(candidate) || candidate <= 0) {
        return DEFAULT_TIMEOUT_MS;
    }
    return Math.floor(candidate);
}
export function resolveFirecrawlMode(options) {
    const candidate = options?.firecrawl;
    if (candidate === 'off' || candidate === 'auto' || candidate === 'always') {
        return candidate;
    }
    return 'auto';
}
export function appendNote(existing, next) {
    if (!next) {
        return existing ?? '';
    }
    if (!existing || existing.length === 0) {
        return next;
    }
    return `${existing}; ${next}`;
}
export function safeHostname(rawUrl) {
    try {
        return new URL(rawUrl).hostname.replace(WWW_PREFIX_PATTERN, '');
    }
    catch {
        return null;
    }
}
export function pickFirstText(candidates) {
    for (const candidate of candidates) {
        const normalized = normalizeCandidate(candidate);
        if (normalized) {
            return normalized;
        }
    }
    return null;
}
export function selectBaseContent(sourceContent, transcriptText) {
    if (!transcriptText) {
        return sourceContent;
    }
    const normalizedTranscript = normalizeForPrompt(transcriptText);
    if (normalizedTranscript.length === 0) {
        return sourceContent;
    }
    return `Transcript:\n${normalizedTranscript}`;
}
export function summarizeTranscript(transcriptText) {
    if (!transcriptText) {
        return { transcriptCharacters: null, transcriptLines: null };
    }
    const transcriptCharacters = transcriptText.length > 0 ? transcriptText.length : null;
    const transcriptLinesRaw = transcriptText
        .split(TRANSCRIPT_LINE_SPLIT_PATTERN)
        .map((line) => line.trim())
        .filter((line) => line.length > 0).length;
    const transcriptLines = transcriptLinesRaw > 0 ? transcriptLinesRaw : null;
    return { transcriptCharacters, transcriptLines };
}
export function ensureTranscriptDiagnostics(resolution, cacheMode) {
    if (resolution.diagnostics) {
        return resolution.diagnostics;
    }
    const hasText = typeof resolution.text === 'string' && resolution.text.length > 0;
    const cacheStatus = cacheMode === 'bypass' ? 'bypassed' : hasText ? 'miss' : 'unknown';
    return {
        cacheMode,
        cacheStatus,
        textProvided: hasText,
        provider: resolution.source,
        attemptedProviders: resolution.source ? [resolution.source] : [],
        notes: cacheMode === 'bypass' ? 'Cache bypass requested' : null,
    };
}
export function finalizeExtractedLinkContent({ url, baseContent, maxCharacters, title, description, siteName, transcriptResolution, diagnostics, }) {
    const normalized = normalizeForPrompt(baseContent);
    const { content, truncated, totalCharacters, wordCount } = typeof maxCharacters === 'number'
        ? applyContentBudget(normalized, maxCharacters)
        : {
            content: normalized,
            truncated: false,
            totalCharacters: normalized.length,
            wordCount: normalized.length > 0
                ? normalized
                    .split(WORD_SPLIT_PATTERN)
                    .map((value) => value.trim())
                    .filter((value) => value.length > 0).length
                : 0,
        };
    const { transcriptCharacters, transcriptLines } = summarizeTranscript(transcriptResolution.text);
    return {
        url,
        title,
        description,
        siteName,
        content,
        truncated,
        totalCharacters,
        wordCount,
        transcriptCharacters,
        transcriptLines,
        transcriptSource: transcriptResolution.source,
        diagnostics,
    };
}
